src/i18n/z_language.erl

```diff
@@ -0,0 +1,1328 @@
+%% @doc Mandatory background read on language tags: [1].
+%%
+%%      Some quotes from [1]:
+%%
+%%          The golden rule when creating language tags is to keep the tag as short as
+%%          possible. Avoid region, script or other subtags except where they add useful
+%%          distinguishing information. For instance, use 'ja' for Japanese and not
+%%          'ja-JP', unless there is a particular reason that you need to say that this is
+%%          Japanese as spoken in Japan, rather than elsewhere.
+%%
+%%          The entries in the registry follow certain conventions with regard to upper
+%%          and lower letter-casing. For example, language tags are lower case, alphabetic
+%%          region subtags are upper case, and script tags begin with an initial capital.
+%%          This is only a convention!
+%%
+%%      Note that we use lower case subtags in subtag identifiers and URLs.
+%%
+%%          Language+extlang combinations are provided to accommodate legacy language tag
+%%          forms, however, there is a single language subtag available for every
+%%          language+extlang combination. That language subtag should be used rather than
+%%          the language+extlang combination, where possible. For example, use 'yue'
+%%          rather than 'zh-yue' for Cantonese, and 'afb' rather than 'ar-afb' for Gulf
+%%          Arabic, if you can.
+
+%%      Language identifiers can have the following forms:
+%%      - language;
+%%      - language-extlang;
+%%      - language-region;
+%%      - language-script;
+%%      It is discouraged to use language-script-region, but it is possible if
+%%      required.
+%%      For a list of language, region and script codes, see [2].
+%%      [1] http://www.w3.org/International/articles/language-tags/
+%%      [2] http://www.iana.org/assignments/language-subtag-registry/language-subtag-registry
+-module(z_language).
+
+-export([
+    default_language/1,
+    is_valid/1,
+    to_language_atom/1,
+    fallback_language/2,
+    english_name/1,
+    is_rtl/1,
+    properties/1,
+    sort_properties/2,
+    all_languages/0,
+    main_languages/0
+]).
+
+-include("zotonic.hrl").
+
+-define(DEFAULT_LANGUAGE, en).
+
+%% @doc Returns the configured default language for this server; if not set, 'en'
+%%      (English).
+-spec default_language(#context{}) -> atom().
+default_language(undefined) -> ?DEFAULT_LANGUAGE;
+default_language(Context) ->
+    z_convert:to_atom(m_config:get_value(i18n, language, ?DEFAULT_LANGUAGE, Context)).
+
+
+%% @doc Check if the language code code is a valid language.
+-spec is_valid(Code::binary() | any()) -> boolean().
+is_valid(Code) when is_binary(Code) ->
+    Languages = all_languages(),
+    proplists:get_value(Code, Languages) /= undefined;
+is_valid(Code) ->
+    is_valid(z_convert:to_binary(Code)).
```

Comment:
Why not accept binaries only instead of doing implicit type conversion? The calling code should convert to binary (if needed).